nightly-examples/examples/echo.rs

```diff
@@ -0,0 +1,119 @@
+use tokio::net::TcpListener;
+
+use std::env;
+use std::error::Error;
+use std::net::SocketAddr;
+
+use tracing::{debug, info, info_span, trace_span, warn};
+use tracing_futures::Instrument;
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn Error>> {
+    let subscriber = tracing_fmt::FmtSubscriber::builder()
+        .with_filter(tracing_fmt::filter::EnvFilter::new("echo=trace"))
+        .finish();
+    tracing::subscriber::set_global_default(subscriber)?;
+
+    // Allow passing an address to listen on as the first argument of this
+    // program, but otherwise we'll just set up our TCP listener on
+    // 127.0.0.1:8080 for connections.
+    let addr = env::args().nth(1).unwrap_or("127.0.0.1:3000".to_string());
+    let addr = addr.parse::<SocketAddr>()?;
+
+    // Next up we create a TCP listener which will listen for incoming
+    // connections. This TCP listener is bound to the address we determined
+    // above and must be associated with an event loop.
+    let mut listener = TcpListener::bind(&addr)?;
+    info!(message = "Listening on", %addr);
+
+    loop {
+        // Asynchronously wait for an inbound socket.
+        let (mut socket, peer_addr) = listener.accept().await?;
+
+        info!(message = "Got connection from", %peer_addr);
+
+        // And this is where much of the magic of this server happens. We
+        // crucially want all clients to make progress concurrently, rather than
+        // blocking one on completion of another. To achieve this we use the
+        // `tokio::spawn` function to execute the work in the background.
+        //
+        // Essentially here we're executing a new task to run concurrently,
+        // which will allow all of our clients to be processed concurrently.
+
+        tokio::spawn(async move {
+            let mut buf = [0; 1024];
+
+            // In a loop, read data from the socket and write the data back.
+            loop {
+                let n: usize = socket
+                    .read(&mut buf)
+                    .map(|bytes| {
+                        if let Ok(n) = bytes {
+                            debug!(message = "read bytes", n);
+                        }
+
+                        bytes
+                    })
+                    .map_err(|error| {
+                        warn!(%error);
+                        error
+                    })
+                    .instrument(trace_span!("read"))
+                    .await
+                    .expect("failed to read data from socket");
+
+                if n == 0 {
+                    return;
+                }
+
+                socket
+                    .write_all(&buf[0..n])
+                    .map(|bytes| {
+                        if let Ok(()) = bytes {
+                            debug!(message = "wrote bytes", n);
+                        }
+
+                        bytes
+                    })
+                    .map_err(|error| {
+                        warn!(%error);
```

Comment:
Again, if we depend on tracing 0.1.6, we don't need `fmt::Display`:
```suggestion
                        warn!(error);
```